### ISSUE TITLE
Instance segmentation: quick but not good enough

### ISSUE THREAD

**Issue by Javacr** [ultralytics/yolov5#10330]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Thanks for your contribution! I have tested the instance segmentation demo, but the results were not good, people's heads always not fully segment, like yolact. I also tried padding, and the same results were presented.
Here are some comparison：
yolov5x-seg result
![Capture](https://user-images.githubusercontent.com/44012990/204452520-0fbb0769-52e4-4400-9641-fcd06009f181.JPG)
yolov7 result
![Capturev7](https://user-images.githubusercontent.com/44012990/204454325-9a52d5de-176e-4e8a-a5c1-faea55eacc8f.JPG)
IDOL result
![Capture_IDOL](https://user-images.githubusercontent.com/44012990/204454602-96c36b2e-2530-4041-8985-40f384b43d93.JPG)

**Comment by JustasBart** [comment]
Hi, I can confirm that there does indeed seem to be a clipping issue with yolov5 segmentations:

![image](https://user-images.githubusercontent.com/40023722/204473831-0ec83905-a2bc-465b-9697-7b2c863ac0a9.png)

![image](https://user-images.githubusercontent.com/40023722/204474570-f2b2f676-0a3b-4ba8-a6e1-a6e6ace041e2.png)

![image](https://user-images.githubusercontent.com/40023722/204474738-8d47729c-2ed3-4942-a062-2a63a41820ab.png)

I'm also seeing what appears to be a shift of the mask to the bottom-right corner and some jagged edges (Although I assume it's just the mask being down scaled as a speed up).

![image](https://user-images.githubusercontent.com/40023722/204475554-6eef7ab3-4d58-487a-bd38-cd9f4b7a67ab.png)

I'm willing to do more training/testing and to try out different settings etc...

Note though that setting the **--mask-ratio 1** does not seem to make any difference when Training/Testing... At least none that I could ever notice...

**Comment by glenn-jocher** [comment]
@JustasBart @Javacr see https://github.com/ultralytics/yolov5/pull/9749 for related PR

**Comment by glenn-jocher** [comment]
@AyushExel @Laughing-q seems like we have a bit of improvement we can make in post-processing/upscaling masks

**Comment by AyushExel** [comment]
@glenn-jocher @Laughing-q @JustasBart @Javacr 
Are we sure that this is an upscaling/post-processing issue? If so then it'll be easier to fix but if not then whatever solution we explore will involve retraining the models.

**Comment by glenn-jocher** [comment]
@AyushExel it seems to be related to upscaling, i.e. see the Sandisk segment boundaries. I tried to work on this in https://github.com/ultralytics/yolov5/pull/9749 didn't find a great solution. Proto 2x linear upsample seems to be a limitation, got good initial results (deterministic) replacing with TransposeConv2d, but this of course would require retraining all models. 

Other than that I feel that the boxes and the segments may always have slight gaps since the box coordinates are continuous and the segments will always be pixelated unless we apply some smoothing/blurring, or unless we just don't threshold and show the full mask within the box on a continuous intensity scale (I discussed this before).

**Comment by Laughing-q** [comment]
@AyushExel @glenn-jocher After some tests, I think I've known why there're jagged edges. And it's about post-processing.
I'll report it here later, once I 100% sure about this and fix it.

**Comment by Laughing-q** [comment]
@JustasBart @Javacr @AyushExel @glenn-jocher 
1. The clipping issue is because we downsampled the bboxes first then cropped masks in function `process_mask`. Because @glenn-jocher found processing this way is faster, and we don't see much gap when we were testing it with 640-size images, if I remember this correctly. But the gap becomes obvious when native size of image is larger.
[https://github.com/ultralytics/yolov5/blob/7f5724ba4b3e421d4c9162742810c52248d06ecd/utils/segment/general.py#L58-L62](https://github.com/ultralytics/yolov5/blob/7f5724ba4b3e421d4c9162742810c52248d06ecd/utils/segment/general.py#L58-L62)
2. The jagged edge is because the upsample in `process_mask` is followed by a `masks.gt_(0.5)`. This is fine if we want to plot in 640-level as `process_mask` outputs 640-level masks. But once we want to plot in higher-level resolution, like native resolution 1920 or even higher, the masks will be upsampled again which the edges are not smooth anymore because the `masks.gt_(0.5)` operation we did before.
[https://github.com/ultralytics/yolov5/blob/7f5724ba4b3e421d4c9162742810c52248d06ecd/utils/plots.py#L137](https://github.com/ultralytics/yolov5/blob/7f5724ba4b3e421d4c9162742810c52248d06ecd/utils/plots.py#L137)

I did a temporary update in `predict.py`.
1. scale the bboxes to nativa-space first then crop the masks, to fix clipping issue.
2. upsample masks to native resolution directly, without `masks.gt(0.5)`, to fix jagged edges.
```python
if len(det):
    # masks = process_mask(proto[i], det[:, 6:], det[:, :4], im.shape[2:], upsample=True)  # HWC
    det[:, :4] = scale_boxes(im.shape[2:], det[:, :4], im0.shape).round()  # rescale boxes to im0 size
    masks = process_mask_native(proto[i], det[:, 6:], det[:, :4], im0.shape[:2])  # HWC
```
and the `process_mask_native`:
```python
def process_mask_native(protos, masks_in, bboxes, dst_shape):
    """
    Crop after upsample.
    proto_out: [mask_dim, mask_h, mask_w]
    out_masks: [n, mask_dim], n is number of masks after nms
    bboxes: [n, 4], n is number of masks after nms
    shape:input_image_size, (h, w)

    return: h, w, n
    """

    c, mh, mw = protos.shape  # CHW
    masks = (masks_in @ protos.float().view(c, -1)).sigmoid().view(-1, mh, mw)
    masks = masks.cpu()
    gain = min(mh / dst_shape[0], mw / dst_shape[1])  # gain  = old / new
    pad = (mw - dst_shape[1] * gain) / 2, (mh - dst_shape[0] * gain) / 2  # wh padding
    top, left = int(pad[1]), int(pad[0])  # y, x
    bottom, right = int(mh - pad[1]), int(mw - pad[0])
    masks = masks[:, top:bottom, left:right]

    masks = F.interpolate(masks[None], dst_shape, mode='bilinear', align_corners=False)[0]  # CHW
    masks = crop_mask(masks, bboxes.cpu())  # CHW
    return masks.gt_(0.5)
```
results:
before:
![image](https://user-images.githubusercontent.com/61612323/204718646-a1ae0d62-64b2-4fa3-b579-3c88e3299941.png)
after the fix:
![image](https://user-images.githubusercontent.com/61612323/204718663-91b5925d-2957-4918-be67-8aaeb6d908b6.png)

**Comment by Laughing-q** [comment]
@glenn-jocher @AyushExel But I haven't profiled the time of this processing. It might be slow, I think I'll test the time then make it work with `--retina-mask` option. I might make a PR for it later, what do you guys think?

**Comment by AyushExel** [comment]
@Laughing-q nice find. Yeah, it'd be interesting to see how much --retina-masks slow down the inference. 
Do you think we should solve both issues only in retina-masks mode? Or does it make sense to solve the clipping issue for default mode also

**Comment by Laughing-q** [comment]
@AyushExel I think just `--retina-mask` mode.

**Comment by JustasBart** [comment]
@Laughing-q Thank you for your work on this, for me personally I live in the C++ environment so for a someone like me speed is never an issue anyway, but the ability to have this extra accuracy would be incredibly useful!

**Comment by glenn-jocher** [comment]
@Javacr @JustasBart good news 😃! Your original issue may now be fixed ✅ in PR #10350 by @Laughing-q. This PR improves and speeds up `--retina-mask` mode during inference, i.e. `python segment/predict.py --retina` will plot smoother and more accurate masks at a slower plotting speed than the default masks.

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by Javacr** [comment]
> @Javacr @JustasBart good news 😃! Your original issue may now be fixed ✅ in PR #10350 by @Laughing-q. This PR improves and speeds up `--retina-mask` mode during inference, i.e. `python segment/predict.py --retina` will plot smoother and more accurate masks at a slower plotting speed than the default masks.
> 
> To receive this update:
> 
> * **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
> * **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
> * **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  [![Run on Gradient](https://camo.githubusercontent.com/c9931a1689c37ab786edd3e1e5f59b9a6f7d097628c4689ce2432563ef884524/68747470733a2f2f6173736574732e706170657273706163652e696f2f696d672f6772616469656e742d62616467652e737667)](https://bit.ly/yolov5-paperspace-notebook) [![Open In Colab](https://camo.githubusercontent.com/84f0493939e0c4de4e6dbe113251b4bfb5353e57134ffd9fcab6b8714514d4d1/68747470733a2f2f636f6c61622e72657365617263682e676f6f676c652e636f6d2f6173736574732f636f6c61622d62616467652e737667)](https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb) [![Open In Kaggle](https://camo.githubusercontent.com/a08ca511178e691ace596a95d334f73cf4ce06e83a5c4a5169b8bb68cac27bef/68747470733a2f2f6b6167676c652e636f6d2f7374617469632f696d616765732f6f70656e2d696e2d6b6167676c652e737667)](https://www.kaggle.com/models/ultralytics/yolov5)
> * **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image [![Docker Pulls](https://camo.githubusercontent.com/280faedaf431e4c0c24fdb30ec00a66d627404e5c4c498210d3f014dd58c2c7e/68747470733a2f2f696d672e736869656c64732e696f2f646f636b65722f70756c6c732f756c7472616c79746963732f796f6c6f76353f6c6f676f3d646f636b6572)](https://hub.docker.com/r/ultralytics/yolov5)
> 
> Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

Thanks for your job again @Laughing-q @glenn-jocher

**Comment by glenn-jocher** [comment]
@Javacr you're welcome! We're always here to help. If you have any other issues or feature requests, don't hesitate to reach out. Happy training with YOLOv5!

**Comment by huzaifaarete** [comment]
Still same problem persists for yolov8L-seg model. Any ideas how to fix it. 
@Laughing-q Sir any ideas will be appreciated.
